nexus-repository-r/src/main/java/org/sonatype/nexus/repository/r/internal/RPackagesBuilderFacetImpl.java

```diff
@@ -0,0 +1,198 @@
+   * posting a {@code RMetadataInvalidationEvent} to the event manager which then triggers the actual metadata
+   * regeneration. Note that the regeneration may not be instantaneous (and that the waiting period is configurable).
+   *
+   * @param basePath The base path of the PACKAGES.gz file to invalidate.
+   */
+  @Override
+  public void invalidateMetadata(final String basePath) {
+    if (!waiting) {
+      eventManager.post(new RMetadataInvalidationEvent(getRepository().getName(), basePath));
+    }
+  }
+
+  /**
+   * Listen for invalidation of the metadata, wait a configured time and then rebuild. The waiting allows throwing away
+   * of subsequent events to reduce the number of rebuilds if multiple archives are being uploaded. This method must NOT
+   * be allowed to process concurrent events as race conditions may result when rebuilding the data.
+   */
+  @Subscribe
+  public void on(final RMetadataInvalidationEvent event) {
+    if (getRepository().getName().equals(event.getRepositoryName())) {
+      try {
+        waiting = true;
+        Thread.sleep(interval);
+      }
+      catch (InterruptedException e) {
+        log.warn("R invalidation thread interrupted on repository {}, proceeding with invalidation",
+            getRepository().getName());
+        Thread.currentThread().interrupt();
+      }
+      waiting = false;
+      log.info("Rebuilding R PACKAGES.gz metadata for repository {}", getRepository().getName());
+      UnitOfWork.begin(getRepository().facet(StorageFacet.class).txSupplier());
+      try {
+        RHostedFacet hostedFacet = getRepository().facet(RHostedFacet.class);
+        hostedFacet.buildAndPutPackagesGz(event.getBasePath());
+      }
+      catch (IOException e) {
+        throw new UncheckedIOException(e);
+      }
+      finally {
+        UnitOfWork.end();
+      }
+    }
+  }
+
+  /**
+   * Returns whether or not an asset event should result in an invalidation request. In order for the event to mandate
+   * that we rebuild metadata, it must both refer to this particular repository, and it also must refer to an uploaded
+   * archive and not to one of the metadata files itself. This helps ensure that we don't end up responding to metadata
+   * changes when rebuilding metadata and end up in a loop.
+   *
+   * @param assetEvent The asset event to process.
+   * @return true if an archive, false if a packages file
+   */
+  private boolean shouldInvalidate(final AssetEvent assetEvent) {
+    return assetEvent.isLocal() &&
+        getRepository().getName().equals(assetEvent.getRepositoryName()) &&
+        isArchiveAssetKind(assetEvent.getAsset());
+  }
+
+  /**
+   * Extracts asset kind from asset and compares to archive.
+   *
+   * @return true is asset kind is archive.
+   */
+  private boolean isArchiveAssetKind(final Asset asset) {
+    return ARCHIVE == RFacetUtils.extractAssetKind(asset);
+  }
+}
```

Comment:
New line after

Reply:
done in fe40919080a00ea9162d0af9be02c414af338e27

Reply:
done in fe40919080a00ea9162d0af9be02c414af338e27